File: ios/Beanshelf/Components.swift
import SwiftUI

/// Full-screen photo viewer — tap anywhere to close. Just looking, no editing.
struct PhotoViewer: View {
    let file: String
    let onDismiss: () -> Void

    @State private var image: UIImage?

    var body: some View {
        ZStack {
            Color.black.opacity(0.95).ignoresSafeArea()
            if let image {
                Image(uiImage: image)
                    .resizable()
                    .scaledToFit()
                    .padding(12)
            }
        }
        .contentShape(Rectangle())
        .onTapGesture(perform: onDismiss)
        .task {
            image = await PhotoStore.loadScaled(file: file, targetWidth: 1600)
        }
    }
}

/// Letter-spaced uppercase metadata label — the bag-label vernacular.
struct Eyebrow: View {
    let text: String
    var color: Color = Palette.dim

    var body: some View {
        Text(text.uppercased())
            .font(Type.labelSmall)
            .kerning(1.8)
            .foregroundStyle(color)
    }
}

/// The signature element: a rotated circular "roast stamp" holding the rating.
/// Double ring + serif numerals on a dark scrim, like an inked stamp on kraft paper.
struct RoastStamp: View {
    let rating: Double
    let size: CGFloat

    var body: some View {
        ZStack {
            Circle()
                .fill(Palette.roast.opacity(0.88))
            Circle()
                .strokeBorder(Palette.stampInk, lineWidth: 2)
                .padding(0.5)
            Circle()
                .strokeBorder(Palette.stampInk.opacity(0.7), lineWidth: 1)
                .padding(5)
            Text(rating <= 0 ? "—" : formatRating(rating))
                .font(.system(size: size * 0.30, weight: .bold, design: .serif))
                .foregroundStyle(Palette.stampInk)
        }
        .frame(width: size, height: size)
        .rotationEffect(.degrees(-12))
    }
}

/// Quarter-step 0–5 rating slider.
struct RatingSlider: View {
    @Binding var value: Double

    var body: some View {
        Slider(value: $value, in: 0...5, step: 0.25)
            .tint(Palette.crema)
    }
}

/// A wooden shelf plank: thin lit top edge, gradient front face, soft shadow below.
/// Drawn full-bleed; bags sit on top of it.
struct ShelfPlank: View {
    var body: some View {
        Canvas { context, size in
            let w = size.width
            let edge: CGFloat = 3
            let face: CGFloat = 18
            // lit top edge
            context.fill(
                Path(CGRect(x: 0, y: 0, width: w, height: edge)),
                with: .color(Palette.plankEdge)
            )
            // front face
            context.fill(
                Path(CGRect(x: 0, y: edge, width: w, height: face)),
                with: .linearGradient(
                    Gradient(colors: [Palette.plankLight, Palette.plankDark]),
                    startPoint: CGPoint(x: 0, y: edge),
                    endPoint: CGPoint(x: 0, y: edge + face)
                )
            )
            // faint wood grain on the face
            for fy in [0.35, 0.62] {
                var line = Path()
                line.move(to: CGPoint(x: 0, y: edge + face * fy))
                line.addLine(to: CGPoint(x: w, y: edge + face * fy))
                context.stroke(line, with: .color(Palette.plankDark.opacity(0.5)), lineWidth: 0.5)
            }
            // shadow cast below the plank
            context.fill(
                Path(CGRect(x: 0, y: edge + face, width: w, height: size.height - edge - face)),
                with: .linearGradient(
                    Gradient(colors: [Color.black.opacity(0.45), .clear]),
                    startPoint: CGPoint(x: 0, y: edge + face),
                    endPoint: CGPoint(x: 0, y: size.height)
                )
            )
        }
        .frame(height: 30)
    }
}

/// Single-select chips that wrap onto multiple lines (Android FilterChip + FlowRow).
struct ChoiceChips: View {
    let options: [String]
    let selected: String
    let onSelect: (String) -> Void

    var body: some View {
        FlowLayout(spacing: 8) {
            ForEach(options, id: \.self) { opt in
                let isSelected = selected == opt
                Button {
                    onSelect(opt)
                } label: {
                    Text(opt)
                        .font(Type.labelLarge)
                        .foregroundStyle(isSelected ? Palette.roast : Palette.dim)
                        .padding(.horizontal, 14)
                        .padding(.vertical, 8)
                        .background(
                            RoundedRectangle(cornerRadius: 8)
                                .fill(isSelected ? Palette.crema : Palette.surface2)
                        )
                        .overlay(
                            RoundedRectangle(cornerRadius: 8)
                                .strokeBorder(
                                    isSelected ? Palette.crema : Palette.dim.opacity(0.35),
                                    lineWidth: 1
                                )
                        )
                }
                .buttonStyle(.plain)
            }
        }
    }
}

/// Minimal wrapping layout (Compose FlowRow equivalent).
struct FlowLayout: Layout {
    var spacing: CGFloat = 8

    func sizeThatFits(proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) -> CGSize {
        let rows = arrange(in: proposal.width ?? .infinity, subviews: subviews)
        let height = rows.last.map { $0.y + $0.height } ?? 0
        return CGSize(width: proposal.width ?? rows.map(\.width).max() ?? 0, height: height)
    }

    func placeSubviews(in bounds: CGRect, proposal: ProposedViewSize, subviews: Subviews, cache: inout ()) {
        let rows = arrange(in: bounds.width, subviews: subviews)
        for row in rows {
            var x = bounds.minX
            for index in row.indices {
                let size = subviews[index].sizeThatFits(.unspecified)
                subviews[index].place(
                    at: CGPoint(x: x, y: bounds.minY + row.y),
                    proposal: ProposedViewSize(size)
                )
                x += size.width + spacing
            }
        }
    }

    private struct Row {
        var indices: [Int] = []
        var y: CGFloat = 0
        var width: CGFloat = 0
        var height: CGFloat = 0
    }

    private func arrange(in maxWidth: CGFloat, subviews: Subviews) -> [Row] {
        var rows: [Row] = []
        var current = Row()
        var x: CGFloat = 0
        var y: CGFloat = 0
        for (i, view) in subviews.enumerated() {
            let size = view.sizeThatFits(.unspecified)
            if x > 0 && x + size.width > maxWidth {
                rows.append(current)
                y += current.height + spacing
                current = Row(y: y)
                x = 0
            }
            current.indices.append(i)
            current.y = y
            current.width = x + size.width
            current.height = max(current.height, size.height)
            x += size.width + spacing
        }
        if !current.indices.isEmpty { rows.append(current) }
        return rows
    }
}

/// Async photo view: loads a downsampled bitmap off the main thread, fills
/// (or fits — cutout PNGs must never be zoom-cropped) its frame, and shows
/// the given placeholder while empty.
struct PhotoImage<Placeholder: View>: View {
    let file: String?
    let targetWidth: CGFloat
    var contentMode: ContentMode = .fill
    var alignment: Alignment = .center
    @ViewBuilder var placeholder: () -> Placeholder

    @State private var image: UIImage?

    var body: some View {
        Color.clear
            .overlay(alignment: alignment) {
                if let image {
                    Image(uiImage: image)
                        .resizable()
                        .aspectRatio(contentMode: contentMode)
                } else {
                    placeholder()
                }
            }
            .clipped()
            .task(id: "\(file ?? "")@\(Int(targetWidth))") {
                image = await PhotoStore.loadScaled(file: file, targetWidth: targetWidth)
            }
    }
}

/// Rounded text-field style matching the Android OutlinedTextField theming.
struct OutlinedField: ViewModifier {
    func body(content: Content) -> some View {
        content
            .tint(Palette.crema)
            .foregroundStyle(Palette.parchment)
            .padding(.horizontal, 14)
            .padding(.vertical, 12)
            .background(
                RoundedRectangle(cornerRadius: 8)
                    .strokeBorder(Palette.dim.opacity(0.4), lineWidth: 1)
            )
    }
}

extension View {
    func outlinedField() -> some View { modifier(OutlinedField()) }
}

File: app/src/main/java/dev/adamsjack/beanshelf/ui/Components.kt
package dev.adamsjack.beanshelf.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import androidx.compose.runtime.getValue
import dev.adamsjack.beanshelf.data.PhotoStore
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Slider
import androidx.compose.material3.SliderDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import dev.adamsjack.beanshelf.model.formatRating

/** Full-screen photo viewer — tap anywhere to close. Just looking, no editing. */
@Composable
fun PhotoViewerDialog(path: String, onDismiss: () -> Unit) {
    val photo by PhotoStore.rememberPhoto(path, targetWidth = 1600)
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false),
    ) {
        Box(
            modifier = Modifier
                .fillMaxSize()
                .background(Color.Black.copy(alpha = 0.95f))
                .clickable(onClick = onDismiss),
            contentAlignment = Alignment.Center,
        ) {
            photo?.let {
                Image(
                    bitmap = it,
                    contentDescription = "Bag photo",
                    contentScale = ContentScale.Fit,
                    modifier = Modifier.fillMaxSize().padding(12.dp),
                )
            }
        }
    }
}

/** Letter-spaced uppercase metadata label — the bag-label vernacular. */
@Composable
fun Eyebrow(text: String, modifier: Modifier = Modifier, color: Color = Dim) {
    Text(
        text = text.uppercase(),
        style = MaterialTheme.typography.labelSmall,
        color = color,
        modifier = modifier,
    )
}

/**
 * The signature element: a rotated circular "roast stamp" holding the rating.
 * Double ring + serif numerals on a dark scrim, like an inked stamp on kraft paper.
 */
@Composable
fun RoastStamp(rating: Float, size: Dp, modifier: Modifier = Modifier) {
    val fontSize = with(size.value) { (this * 0.30f).sp }
    Box(
        modifier = modifier
            .size(size)
            .graphicsLayer { rotationZ = -12f },
        contentAlignment = Alignment.Center,
    ) {
        Canvas(Modifier.size(size)) {
            val r = this.size.minDimension / 2f
            drawCircle(color = Roast.copy(alpha = 0.88f), radius = r)
            drawCircle(color = StampInk, radius = r - 1.5.dp.toPx(), style = Stroke(width = 2.dp.toPx()))
            drawCircle(color = StampInk.copy(alpha = 0.7f), radius = r - 5.5.dp.toPx(), style = Stroke(width = 1.dp.toPx()))
        }
        Text(
            text = if (rating <= 0f) "—" else formatRating(rating),
            color = StampInk,
            fontFamily = FontFamily.Serif,
            fontWeight = FontWeight.Bold,
            fontSize = fontSize,
        )
    }
}

/** Quarter-step 0–5 rating slider with a live stamp preview beside it. */
@Composable
fun RatingSlider(value: Float, onChange: (Float) -> Unit, modifier: Modifier = Modifier) {
    Slider(
        value = value,
        onValueChange = { onChange((it * 4).toInt() / 4f) },
        valueRange = 0f..5f,
        steps = 19,
        modifier = modifier,
        colors = SliderDefaults.colors(
            thumbColor = Crema,
            activeTrackColor = Crema,
            inactiveTrackColor = SurfaceHigh,
            activeTickColor = Roast.copy(alpha = 0.3f),
            inactiveTickColor = Dim.copy(alpha = 0.2f),
        ),
    )
}

/**
 * A wooden shelf plank: thin lit top edge, gradient front face, soft shadow below.
 * Drawn full-bleed; bags sit on top of it.
 */
@Composable
fun ShelfPlank(modifier: Modifier = Modifier) {
    Canvas(modifier.fillMaxWidth().height(30.dp)) {
        val w = this.size.width
        val edge = 3.dp.toPx()
        val face = 18.dp.toPx()
        // lit top edge
        drawRect(color = PlankEdge, size = androidx.compose.ui.geometry.Size(w, edge))
        // front face
        drawRect(
            brush = Brush.verticalGradient(
                colors = listOf(PlankLight, PlankDark),
                startY = edge,
                endY = edge + face,
            ),
            topLeft = Offset(0f, edge),
            size = androidx.compose.ui.geometry.Size(w, face),
        )
        // faint wood grain on the face
        val grainY = listOf(0.35f, 0.62f)
        grainY.forEach { fy ->
            drawLine(
                color = PlankDark.copy(alpha = 0.5f),
                start = Offset(0f, edge + face * fy),
                end = Offset(w, edge + face * fy),
                strokeWidth = 1f,
            )
        }
        // shadow cast below the plank
        drawRect(
            brush = Brush.verticalGradient(
                colors = listOf(Color.Black.copy(alpha = 0.45f), Color.Transparent),
                startY = edge + face,
                endY = this.size.height,
            ),
            topLeft = Offset(0f, edge + face),
            size = androidx.compose.ui.geometry.Size(w, this.size.height - edge - face),
        )
    }
}
